src/mame/drivers/magicard.cpp

```diff
@@ -350,313 +571,656 @@ TODO: check this register,doesn't seem to be 100% correct.
+	PORT_DIPNAME( 0x02, 0x02, "Remote/Keyboard" )			PORT_DIPLOCATION("DIP 1:7")
+	PORT_DIPSETTING(    0x02, "Remote Switch" )
+	PORT_DIPSETTING(    0x00, "Keyboard" )
+	PORT_DIPNAME( 0x01, 0x01, "Keyboard Test" )				PORT_DIPLOCATION("DIP 1:8")
+	PORT_DIPSETTING(    0x01, DEF_STR( Off ) )
 	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
-	PORT_DIPNAME( 0x08, 0x08, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x08, DEF_STR( Off ) )
+
+	PORT_BIT( 0x8000, IP_ACTIVE_LOW, IPT_BUTTON7 )     PORT_NAME("Hopper Full")   PORT_CODE(KEYCODE_A)  PORT_TOGGLE
+	PORT_BIT( 0x4000, IP_ACTIVE_LOW, IPT_UNUSED )      // PORT_NAME("Reserve In 9")
+	PORT_BIT( 0x2000, IP_ACTIVE_LOW, IPT_UNUSED )      // PORT_NAME("Reserve In 8")
+	PORT_BIT( 0x1000, IP_ACTIVE_LOW, IPT_UNUSED )      // PORT_NAME("Reserve In 7")
+	PORT_BIT( 0x0800, IP_ACTIVE_LOW, IPT_UNUSED )      // PORT_NAME("Reserve In 6")
+	PORT_BIT( 0x0400, IP_ACTIVE_LOW, IPT_UNUSED )      // PORT_NAME("Reserve In 5")
+	PORT_BIT( 0x0200, IP_ACTIVE_LOW, IPT_GAMBLE_DOOR ) PORT_NAME("Door Switch")   PORT_TOGGLE
+	PORT_BIT( 0x0100, IP_ACTIVE_LOW, IPT_BUTTON13 )    PORT_NAME("Clear Credit")  PORT_CODE(KEYCODE_J)
+
+INPUT_PORTS_END
+
+static INPUT_PORTS_START( puzzleme )
+	PORT_START("SW0")
+
+	PORT_BIT( 0x0001, IP_ACTIVE_LOW, IPT_COIN1 )
+	PORT_BIT( 0x0002, IP_ACTIVE_LOW, IPT_COIN3 ) PORT_NAME("Remote")
+	PORT_BIT( 0x0004, IP_ACTIVE_LOW, IPT_BUTTON5 )
+	PORT_BIT( 0x0008, IP_ACTIVE_LOW, IPT_COIN2 )
+	PORT_BIT( 0x0010, IP_ACTIVE_LOW, IPT_JOYSTICK_LEFT )
+	PORT_BIT( 0x0020, IP_ACTIVE_LOW, IPT_BUTTON1 )  PORT_NAME("Clear")
+	PORT_BIT( 0x0040, IP_ACTIVE_LOW, IPT_BUTTON2 )  PORT_NAME("Show All Book Keeping")
+	PORT_BIT( 0x0080, IP_ACTIVE_LOW, IPT_START1 )
+	PORT_BIT( 0x0100, IP_ACTIVE_LOW, IPT_BUTTON3 )  PORT_NAME("Rental Book Keeping") PORT_CODE(KEYCODE_R)
+	PORT_BIT( 0x0200, IP_ACTIVE_LOW, IPT_BUTTON4 )  PORT_NAME("Owner Book Keeping") PORT_CODE(KEYCODE_O)
+	PORT_BIT( 0x0400, IP_ACTIVE_LOW, IPT_JOYSTICK_DOWN )
+	PORT_BIT( 0x0800, IP_ACTIVE_LOW, IPT_JOYSTICK_RIGHT )
+	PORT_BIT( 0x1000, IP_ACTIVE_LOW, IPT_JOYSTICK_UP )
+	PORT_BIT( 0x2000, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x4000, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x8000, IP_ACTIVE_LOW, IPT_UNUSED )
+
+	// the dip switches don't appear to be used for anything
+	PORT_START("SW1")
+	PORT_BIT( 0x0001, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x0002, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x0004, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x0008, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x0010, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x0020, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x0040, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x0080, IP_ACTIVE_LOW, IPT_UNUSED )
+
+	PORT_BIT( 0x8000, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x4000, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x2000, IP_ACTIVE_LOW, IPT_UNUSED )
+	PORT_BIT( 0x1000, IP_ACTIVE_LOW, IPT_BUTTON6 )  PORT_NAME("Easy 1")		PORT_TOGGLE
+	PORT_BIT( 0x0800, IP_ACTIVE_LOW, IPT_BUTTON7 )  PORT_NAME("Easy 2")		PORT_TOGGLE
```

Comment:
also dipswitches should always be mapped as dipswitches, not as PORT_BIT( 0x0001, IP_ACTIVE_LOW, IPT_UNUSED )

it's unimportant that they have no functionality, if they exist they need to be there

Reply:
OK, makes sense that things have to be defined if they exist but not used. Didn't realise though.

I don't know whether there are two full banks of switches on this or not. All I had to go on was the ascii artwork in the driver file and that doesn't show any dil switches at all on the puzzleme drawing :(
Can you confirm whether it's two full  8-bit banks or whether it's one bank with the other 8 bits being normal inputs (similar to other boards) ?

Purely for interest, if photos of original boards are available is there a reason to fill source files with ascii art rather than just link to a proper image ? Mamedevs must get sent a huge amount of useful info which I assume is archived somewhere, can that be accessed by others ?